### DISCUSSION TITLE
Setting x-message-ttl Exception

### DISCUSSION THREAD

**Discussion by ZedZipDev** [rabbitmq/rabbitmq-dotnet-client#1115] [General]
I have an existing Queue with messages and want to change the Queues TTL. 

Do it:
```
var args = new Dictionary<string, object>();
args.Add("x-message-ttl", 20000);
model.QueueDeclare("DoBeDo", false, false, false, args);

```

But receive: 
```
The AMQP operation was interrupted: AMQP close-reason, initiated by Peer, code=406, text='PRECONDITION_FAILED - 
inequivalent arg 'x-message-ttl' for queue 'DoBeDo' in vhost '/': 
received the value '20000' of type 'signedint' but current is none', classId=50, methodId=10

```

What is wrong? Is it possible to change the TTL for the existing Queue with messages?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
You cannot modify properties of a previously declared queue: they are [immutable and checked for equivalence](https://rabbitmq.com/queues.html#property-equivalence). Either delete the queue and re-declare it, or delete it, redeclare and [use a policy instead](https://rabbitmq.com/parameters.html#policies).

**Reply by ZedZipDev** [reply]
Thank you.
Probably the doc should be corrected here as you wrote: 

https://www.rabbitmq.com/ttl.html#per-message-ttl-caveats

```
It is possible to apply a message TTL policy to a queue which already has messages in it but this involves some caveats.

The original expiry time of a message is preserved if it is requeued (for example due to the use of an AMQP method that features a requeue parameter, or due to a channel closure).
```

**Reply by ZedZipDev** [reply]
And yet another thing:
If I do it for the existing queue I catch the Exception but the channel closed and I cannot use it anymore. 
IMHO it will be better to generate exception during the TTL setting for existing queue and continue to work

```
           var args = new Dictionary<string, object>();
            args.Add("x-message-ttl", 30000);
            try
            {
               channel.QueueDeclare(QueName, true, false, false, args);
            }
            catch(Exception x)
            {
                Console.WriteLine(" Cannot change TTL: {0}", x.Message);//<--- after  it the channel is closed.
            }
```

**Reply by sungam3r** [reply]
> during the TTL **setting**

What is it?

**Reply by sungam3r** [reply]
>  the channel closed and I cannot use it anymore

It is intended behavior due to AMQP spec.

**Reply by ZedZipDev** [reply]
"What is it?"
It is :
> var args = new Dictionary<string, object>();
>             **args.Add("x-message-ttl", 30000);**
>             try
>             {
>                **channel.QueueDeclare(QueName, true, false, false, args);**

**Reply by sungam3r** [reply]
The exception is generated when the queue is declared/redeclared.

>  for existing queue

How to know if queue exists or not ?

**Reply by ZedZipDev** [reply]
I run the dashboard and see all queues including this queue

**Reply by ZedZipDev** [reply]
@sungam3r  but channel closed too in this case and I cannot work with it. That is the problem.

**Comment by ZedZipDev** [comment]
How to check if queue already exists? 
When I try in my c# app to change TTL for the queue I catch an exception and channel is closed. 
How to check if queue already exists to prevent this situation? i.e. if queue do not exist I create it with TTL and other parameters else I do not try to create it.

**Reply by sungam3r** [reply]
https://www.rabbitmq.com/api-guide.html#passive-declaration

**Reply by ZedZipDev** [reply]
I use this approach. 
There is my function:
```
 public static bool RMQQueueExists(IModel channel, string QueName)
        {
            bool b = false;
            try
            {
                var xxOk = channel.QueueDeclarePassive(QueName);
                b = true;
            }
            catch (Exception x)
            {
                Console.WriteLine("Queue {0} does not exist", x.Message); 
              // and channel.IsClosed  here!
            }
            return b;
        }

```
In case of queue does not exist (exception) the channel closed and cannot be used anymore.

Please correct me what is wrong?
